benchexec/tools/metaval.py

```diff
@@ -0,0 +1,82 @@
+"""
+BenchExec is a framework for reliable benchmarking.
+This file is part of BenchExec.
+
+Copyright (C) 2007-2019  Dirk Beyer
+All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import argparse
+import benchexec.util as util
+import benchexec.tools.template
+import benchexec.result as result
+import os
+
+
+class Tool(benchexec.tools.template.BaseTool):
+
+    TOOL_TO_PATH_MAP = {"cpachecker": "CPAchecker", "esbmc": "esbmc"}
+    REQUIRED_PATHS = list(TOOL_TO_PATH_MAP.values())
+
+    def executable(self):
+        return util.find_executable("metaval.sh")
+
+    def name(self):
+        return "metaval"
+
+    def determine_result(self, returncode, returnsignal, output, isTimeout):
+        if not hasattr(self, "wrappedTool"):
+            return "METAVAL ERROR"
+        return self.wrappedTool.determine_result(
+            returncode, returnsignal, output, isTimeout
+        )
+
+    def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}):
+        parser = argparse.ArgumentParser(add_help=False)
+        parser.add_argument("--witness")
+        parser.add_argument("--metaval")
+        (knownargs, options) = parser.parse_known_args(options)
+        self.verifierName = knownargs.metaval.lower()
+        self.witnessName = knownargs.witness
+        self.wrappedTool = __import__("benchexec.tools." + self.verifierName, fromlist=["Tool"]).Tool()
+
+        if hasattr(self, "wrappedTool"):
+            oldcwd = os.getcwd()
+            os.chdir(os.path.join(oldcwd, self.TOOL_TO_PATH_MAP[self.verifierName]))
```

Comment:
Hm, unfortunately, this won't work. For example, if two threads use this tool-info module at the same time, they will overwrite their respective changes to the current directory. (Even if this was not a problem, the second `chdir` would need to be in a `finally` block.)

Maybe the best is to consider how to make CPAchecker's tool-info module find `cpa.sh` in your case. Manipulating `os.environ["PATH"]` would be an option, but has the same problem. Some changes to CPAchecker's tool-info module or `util.find_executable()` maybe?

A last possibility would be to add an optional parameter with a base dir to CPAchecker's `executable()` method. Metaval's module would then attempt to call the method with the parameter and fall back to calling it without. But maybe there are better options.

Reply:
> Manipulating os.environ["PATH"] would be an option, but has the same problem

Yes, I did this before. That was why I added the "scripts" in the `TOOL_TO_PATH_MAP`.
What about locking and thus making this transaction atomic?

> A last possibility would be to add an optional parameter with a base dir to CPAchecker's executable() method

The problem is not just CPAchecker, potentially every tool ...

Reply:
By the way:
> For example, if two threads use this tool-info module at the same time, they will overwrite their respective changes to the current directory.

Why isn't it designed in a way that the two threads use different tool instances?

Reply:
> What about locking and thus making this transaction atomic?

I guess that could work. The lock should cover only the smallest amount of possible code (calling `executable()`, not `cmdline()`). It can be combined with the lock that covers the loading of the tool-info module, if the result of `executable()` is also stored in a field.

But did you actually test this? The result of doing `chdir` will be that the paths returned by `executable()` and `cmdline()` are relative to that directory, and this shouldn't work for the actual run, which is executed from a different directory. You could specify a working directory for the run (in the respective method), but this would make Metaval more difficult for users to execute manually, I guess?

> Why isn't it designed in a way that the two threads use different tool instances?

Why should it be? Tool-info modules are supposed to be stateless anyway. And it just fits naturally into BenchExec's architecture that way.

Reply:
> But did you actually test this? The result of doing chdir will be that the paths returned by executable() and cmdline() are relative to that directory, and this shouldn't work for the actual run, which is executed from a different directory.

Yes, I tested this and it works nicely. This is why `--verifier` is passed to the tool, such that it can figure out relative to which folder the paths are specified. This way everything is identical to as if I would call the tool directly. Some tools like esbmc do not support being called from outside of their folder (`esbmc-wrapper.py` hardcoded `the path to the executable as "./esbmc"` for example)

> You could specify a working directory for the run (in the respective method), but this would make Metaval more difficult for users to execute manually, I guess?

I guess this is essentially what the `--verifier` option now does (though the resolving of toolname to verifier dir is done by the tool, as to not burden the user with that choice).

> Why should it be? Tool-info modules are supposed to be stateless anyway. And it just fits naturally into BenchExec's architecture that way.

I am not talking about the module, but the instance of the subclass of the class `Tool`

Reply:
> tool-info modules are supposed to be stateless anyway.

Hmm, then instead of having a state in `self.wrappedTool` for `determine_result`, I could put the tool name at the start of the ouput and read this in again once determine_result is called :thinking:.

Would that be an improvement in your opinion?

Reply:
As long as you do not require different validators per task, the existing solution is ok for me. (It is not mutable state, just lazily initialized constant state.) For a better general-purpose solution there is #166.